```{r, show=FALSE, warnings=FALSE}
library(tidyverse)
library(Seurat)
library(wesanderson)
`%notin%` <- negate(`%in%`)
source('functions/deFunctions.R')
```

## Load classified data
```{r}
mb231 <- readRDS('./data/rds/mb231_colors_cellclass_hclust_nb-barassigned_ccclustered_precc_singlets.rds')
h1806 <- readRDS('./data/rds/h1806_colors_cellclass_hclust_nb-barassigned_ccclustered_precc_singlets.rds')
```

## Subset data
Only keep clones in the fusion-enriched subset with at least `n_cell_min` cells present. For this differential expression analysis, we will compare fusion cells in cluster F versus all other fusion cells
```{r}
fuse_ctrl_mincells_h1806 <- subset_on_fusion_v_ctrl(h1806, n_cell_min=5)
fc_h1806 <- fuse_ctrl_mincells_h1806[[1]]
fc_h1806 <- subset(fc_h1806, cell_class == 'fusion')
fc_h1806@meta.data <- fc_h1806@meta.data %>% mutate(cell_class = case_when(clust == 'F' ~ 'Fclust', .default = 'other'))
Idents(fc_h1806) <- "cell_class"

fuse_ctrl_mincells_mb231 <- subset_on_fusion_v_ctrl(mb231, n_cell_min=5)
fc_mb231 <- fuse_ctrl_mincells_mb231[[1]]
fc_mb231 <- subset(fc_mb231, cell_class == 'fusion')
fc_mb231@meta.data <- fc_mb231@meta.data %>% mutate(cell_class = case_when(clust == 'F' ~ 'Fclust', .default = 'other'))
Idents(fc_mb231) <- "cell_class"
```

## Inspect clonal composition of subsetted data
```{r}
plot_clonal_composition_byclass(fc_h1806, class_levels = c('other','Fclust')) + ggtitle('HCC1806')
plot_clonal_composition_byclass(fc_mb231, class_levels = c('other','Fclust')) + ggtitle('MDA-MB-231')
```


## Ask if there are any cell cycle differences between clusters
```{r}
source('functions/plot_cellCycle_by_cluster.R')
fig_path <- "./plots/cell_cycle_phase/"

fs = 10
ls = 3

hcc <- plot_cellCycle_by_cluster(h1806, cellline_name='HCC1806', subtext = 'all cells in each cluster', 
                          perc.label.size = ls, font.size = fs, label_vShift = 0.05)
mcc <- plot_cellCycle_by_cluster(mb231, cellline_name='MDA-MB-231', subtext = 'all cells in each cluster',
                          perc.label.size = ls, font.size = fs, label_vShift = 0.07) 

hfcc <- plot_cellCycle_by_cluster(fc_h1806, cellline_name='HCC1806', subtext = 'fusion cells in each cluster',
                          perc.label.size = ls, font.size = fs, label_vShift = 0.05)
mfcc <- plot_cellCycle_by_cluster(fc_mb231, cellline_name='MDA-MB-231', subtext = 'fusion cells in each cluster',
                          perc.label.size = ls, font.size = fs, label_vShift = 0.05) 
hcc
hfcc
mcc
mfcc

# SaveFigure(hcc, 'h1806_cell_cycle_phase_all_cells', type = "png", width=8, height=4, res=600)
# SaveFigure(mcc, 'mb231_cell_cycle_phase_all_cells', type = "png", width=8, height=4, res=600)
# SaveFigure(hfcc, 'h1806_cell_cycle_phase_fusion_cells', type = "png", width=8, height=4, res=600)
# SaveFigure(mfcc + scale_x_discrete(limits=LETTERS[1:7]), 'mb231_cell_cycle_phase_fusion_cells', type = "png", width=8, height=4, res=600)


fc_mb231_Fvs <- fc_mb231
fc_mb231_Fvs@meta.data <- fc_mb231_Fvs@meta.data %>% mutate(clust = case_when(clust=='F' ~ 'F', .default='other'))
mF_only <- plot_cellCycle_by_cluster(fc_mb231_Fvs, cellline_name='MDA-MB-231', subtext = 'fusion cells in cluster F vs. other clusters',
                          perc.label.size = ls, font.size = fs, label_vShift = 0.05) 

fc_h1806_Fvs <- fc_h1806
fc_h1806_Fvs@meta.data <- fc_h1806_Fvs@meta.data %>% mutate(clust = case_when(clust=='F' ~ 'F', .default='other'))
hF_only <- plot_cellCycle_by_cluster(fc_h1806_Fvs, cellline_name='HCC1806', subtext = 'fusion cells in cluster F vs. other clusters',
                          perc.label.size = ls, font.size = fs, label_vShift = 0.05) 


SaveFigure(mF_only, 'mb231_cell_cycle_phase_fusion_cells_clustF', type = "png", width=8, height=4, res=600)
SaveFigure(hF_only, 'h1806_cell_cycle_phase_fusion_cells_clustF', type = "png", width=8, height=4, res=600)

mF_only
hF_only
```


## Perform statistical analysis of gene expression differences
Above we see that the clonal composition is skewed, especially so in the MDA-MB-231 cells. To account for underlying clonal composition, we will used a mixed model to perform differential expression analysis, setting cell barcode as the random effect (i.e. the subject id). MAST (https://github.com/RGLab/MAST) is a package which allows for this analysis and more complex models, but is slow. NEBULA (https://github.com/lhe17/nebula) has a much faster implementation for simple mixed models when accounting for a single random effect. We will use NEBULA.

```{r}
sample_name = 'fusion_in_clustF_vs_other'
min_pct =0.1
test_group = 'other'


# run mixed model for DE analysis
h1806_neb <- run_nebula_de(fc_h1806, method='LN', min.pct=min_pct, n_cores=16, min.cpc=0.01, sid = 'barpairid', group = c('cell_class','clust'))
mb231_neb <- run_nebula_de(fc_mb231, method='LN', min.pct=min_pct, n_cores=16, min.cpc=0.01, sid = 'barpairid', group = c('cell_class','clust'))

write.table(h1806_neb[[1]], file=paste0('outdata/de_results/nebula/nebula_',min_pct,'pct_h1806_',sample_name,'_raw.tsv'),
            quote=FALSE, sep='\t', col.names = TRUE)
write.table(mb231_neb[[1]], file=paste0('outdata/de_results/nebula/nebula_',min_pct,'pct_mb231_',sample_name,'_raw.tsv'),
            quote=FALSE, sep='\t', col.names = TRUE)

capture.output(h1806_neb, file=paste0('outdata/de_results/nebula/nebula_qc/nebula_',min_pct,'pct_h1806_',sample_name,'.tsv'))
capture.output(mb231_neb, file=paste0('outdata/de_results/nebula/nebula_qc/nebula_',min_pct,'pct_mb231_',sample_name,'.tsv'))

# calculate Pearson residuals
h1806_pres <- get_pearson_nebula(h1806_neb, fc_h1806, method='LN', min.pct=0.1, n_cores=16, min.cpc=0.01, sid = 'barpairid', group = 'cell_class')
mb231_pres <- get_pearson_nebula(mb231_neb, fc_mb231, method='LN', min.pct=0.1, n_cores=16, min.cpc=0.01, sid = 'barpairid', group = 'cell_class')

capture.output(h1806_pres, file=paste0('outdata/de_results/nebula/nebula_qc/nebula_',min_pct,'pct_h1806_',sample_name,'_pearson_residuals.tsv'))
capture.output(mb231_pres, file=paste0('outdata/de_results/nebula/nebula_qc/nebula_',min_pct,'pct_mb231_',sample_name,'_pearson_residuals.tsv'))

# format outputs for downstream analysis
h1806_neb_format <- format_neb_outs(h1806_neb, testgroup = test_group) #%>% mutate(log2FC = -log2FC) # test_group was flipped
mb231_neb_format <- format_neb_outs(mb231_neb, testgroup = test_group) #%>% mutate(log2FC = -log2FC) # test_group was flipped

write.table(h1806_neb_format, file=paste0('outdata/de_results/nebula/nebula_',min_pct,'pct_h1806_',sample_name,'.tsv'),
            quote=FALSE, sep='\t', col.names = TRUE)
write.table(mb231_neb_format, file=paste0('outdata/de_results/nebula/nebula_',min_pct,'pct_mb231_',sample_name,'.tsv'),
            quote=FALSE, sep='\t', col.names = TRUE)
```


```{r}
sessionInfo()
```

