```{r, show=FALSE, warnings=FALSE}
library(tidyverse)
library(Seurat)
library(wesanderson)
`%notin%` <- negate(`%in%`)
source('functions/deFunctions.R')
```

## Load classified data
```{r}
mb231 <- readRDS('./data/rds/mb231_cellclass_hclust_nb-barassigned_ccclustered_precc_singlets.rds')
h1806 <- readRDS('./data/rds/h1806_cellclass_hclust_nb-barassigned_ccclustered_precc_singlets.rds')
```

## Subset data
Only keep clones in the fusion-enriched or control samples with at least `n_cell_min` cells present. For this differential expression analysis, we will only use cells found in the control sorted samples or cells in the fusion sorted samples to control for expression differences due to cell sorting. 
```{r}
fuse_ctrl_mincells_h1806 <- subset_on_fusion_v_ctrl(h1806, n_cell_min=5)
fc_h1806 <- fuse_ctrl_mincells_h1806[[1]]
Idents(fc_h1806) <- "cell_class"

fuse_ctrl_mincells_mb231 <- subset_on_fusion_v_ctrl(mb231, n_cell_min=5)
fc_mb231 <- fuse_ctrl_mincells_mb231[[1]]
Idents(fc_mb231) <- "cell_class"
```

## Inspect clonal composition of subsetted data
```{r}
plot_clonal_composition_byclass(fc_h1806, class_levels = c('control','fusion')) + ggtitle('HCC1806')
plot_clonal_composition_byclass(fc_mb231, class_levels = c('control','fusion')) + ggtitle('MDA-MB-231')
```

## Perform statistical analysis of gene expression differences
Above we see that the clonal composition is skewed, especially so in the MDA-MB-231 cells. To account for underlying clonal composition, we will used a mixed model to perform differential expression analysis, setting cell barcode as the random effect (i.e. the subject id). MAST (https://github.com/RGLab/MAST) is a package which allows for this analysis and more complex models, but is slow. NEBULA (https://github.com/lhe17/nebula) has a much faster implementation for simple mixed models when accounting for a single random effect. We will use NEBULA.
```{r}
# run mixed model for DE analysis
h1806_neb <- run_nebula_de(fc_h1806, method='LN', min.pct=0.1, n_cores=16, min.cpc=0.01, sid = 'barpairid', group = 'cell_class')
mb231_neb <- run_nebula_de(fc_mb231, method='LN', min.pct=0.1, n_cores=16, min.cpc=0.01, sid = 'barpairid', group = 'cell_class')

write.table(h1806_neb[[1]], file='outdata/de_results/nebula/nebula_0.1pct_h1806_fusion_v_control_raw.tsv', quote=FALSE, sep='\t', col.names = TRUE)
write.table(mb231_neb[[1]], file='outdata/de_results/nebula/nebula_0.1pct_mb231_fusion_v_control_raw.tsv', quote=FALSE, sep='\t', col.names = TRUE)

capture.output(h1806_neb, file='outdata/de_results/nebula/nebula_qc/nebula_0.1pct_h1806_fusion_v_control.tsv')
capture.output(mb231_neb, file='outdata/de_results/nebula/nebula_qc/nebula_0.1pct_mb231_fusion_v_control.tsv')

# calculate Pearson residuals
h1806_pres <- get_pearson_nebula(h1806_neb, fc_h1806, method='LN', min.pct=0.1, n_cores=16, min.cpc=0.01, sid = 'barpairid', group = 'cell_class')
mb231_pres <- get_pearson_nebula(mb231_neb, fc_mb231, method='LN', min.pct=0.1, n_cores=16, min.cpc=0.01, sid = 'barpairid', group = 'cell_class')

capture.output(h1806_pres, file='outdata/de_results/nebula/nebula_qc/nebula_0.1pct_h1806_fusion_v_control_pearson_residuals.tsv')
capture.output(mb231_pres, file='outdata/de_results/nebula/nebula_qc/nebula_0.1pct_mb231_fusion_v_control_pearson_residuals.tsv')

# format outputs for downstream analysis
h1806_neb_format <- format_neb_outs(h1806_neb, testgroup = 'fusion')
mb231_neb_format <- format_neb_outs(mb231_neb, testgroup = 'fusion')

write.table(h1806_neb_format, file='outdata/de_results/nebula/nebula_0.1pct_h1806_fusion_v_control.tsv', quote=FALSE, sep='\t', col.names = TRUE)
write.table(mb231_neb_format, file='outdata/de_results/nebula/nebula_0.1pct_mb231_fusion_v_control.tsv', quote=FALSE, sep='\t', col.names = TRUE)
```


```{r}
sessionInfo()
```

